src/lib.rs

```diff
@@ -1 +1,406 @@
+pub struct Client {
+    #[serde(rename = "client")]
+    id: u16,
+    available: Decimal,
+    held: Decimal,
+    total: Decimal,
+    locked: bool,
+}
+
+impl Client {
+    pub fn new(id: u16) -> Self {
+        Self {
+            id,
+            ..Default::default()
+        }
+    }
+}
+
+fn process_records(records: &Records) -> Result<Vec<Client>> {
+    let mut clients = BTreeMap::new();
+    for record in records.0.values() {
+        // We take a copy of the `Client` and overwrite it in the map later to ensure atomicity of
+        // the transaction.
+        let mut client = *clients
+            .entry(record.client)
+            .or_insert_with(|| Client::new(record.client));
+        // TODO - remove
+        // println!("record: {:?}", record);
+        // TODO - what should we do if an account is already frozen?
+        match record.record_type {
+            RecordType::Deposit => {
+                client.available += record.amount;
+                client.total += record.amount;
+            }
+            RecordType::Withdrawal => {
+                ensure!(
+                    client.available >= record.amount,
+                    "Withdrawal failed. Available funds insufficient."
+                );
+                client.available -= record.amount;
+                client.total -= record.amount;
+            }
+            RecordType::Dispute => {
+                let disputed_record = records.0.get(&Id::Tx(record.tx)).context(format!(
+                    "Disputed record tx {} could not be found",
+                    record.tx
+                ))?;
+                // TODO - what happens if available is less than disputed amount?
+                client.available -= disputed_record.amount;
+                client.held += disputed_record.amount;
+            }
+            RecordType::Resolve => {
+                let resolved_record = records.0.get(&Id::Tx(record.tx)).context(format!(
+                    "Resolved record tx {} could not be found",
+                    record.tx
+                ))?;
+                // TODO - what happens if held is less than resolved amount?
+                client.available += resolved_record.amount;
+                client.held -= resolved_record.amount;
+            }
+            RecordType::Chargeback => {
+                let chargeback_record = records.0.get(&Id::Tx(record.tx)).context(format!(
+                    "Chargeback record tx {} could not be found",
+                    record.tx
+                ))?;
+                // TODO - what happens if available/held are less than chargeback amount?
+                client.available -= chargeback_record.amount;
+                client.held -= chargeback_record.amount;
+                client.locked = true;
+            }
+        }
+
+        // TODO - remove this debug aid
+        // println!("{:?}", client);
+
+        // Atomically update the map with our transaction by copying over the value in the map.
```

Comment:
```suggestion
        // Atomically update the map with our transaction by copying over the value in the map. If this were a real database
        // we would need to verify that the data in the database has not changed since we first accessed it. This can be done with object version numbers or database transactions. (Watch out for database transactions though, they can get dangerous!)
```